### ISSUE TITLE
Feature Request - Attachment type and File Manager folder structures

### ISSUE THREAD

**Issue by dottenbr** [frappe/erpnext#4648]
I suggest every DocType has configurable attachment types that are in se selection list when a user uploads a new file.This would allow a clean structure within the DocType. Furthermore it would be ideal if the filemanager had a folder structure rather than a flat hirarchy. So for example if I uploaded a "certificate" type attachment into a supplier it should be stored under file manager folder structure: Supplier > Supplier Name > Certificate > {File}.

**Comment by rmehta** [comment]
This was implemented in the first cut, but we scrapped it because the attachments are already discoverable via the document

**Comment by dottenbr** [comment]
Yes absolutely @rmehta we are using ERPNext to document our products and to upload the customer specific manufacturing data. So when in an Item we would like to upload the "Customer Data" as one file type, "Amanded Documentation" or "Quality Requirements" or whatever we think of creating as specific upload types in doctypes. This allows us to use ERPNext as a data container but to at the same time store the information in a basically structured way rather than just dumping all data in.

The storing in the File Manager in a 100% flat hirarchy just plainly drives me insane and I dont see why we wouldnt follow the app > doctype > document > uploadtype (new) scheme to ensure data is stored in a ordered way. We will have thousands of files stored.

**Comment by dottenbr** [comment]
@rmehta and @anandpdoshi on another not I think the idea of "private" attachments is quite dangerous. I think you need to ensure that every attachment is "private" by default unless a user is allowed to upload a "pubic" attachment. Consider people are uploading invoices from suppliers or customer data, if someone made that "non private" or "public" that could be disastrous.

Also how do you ensure that a user who is not allowed to open Human Resources or another module can't open attachments of that module? I don't want all ERPNext users to be able to read customer data.

**Comment by dottenbr** [comment]
Any Way i can convince you to do this? We are starting to use the system a lot more and this is really not ideal at the moment. I think the structure would be perfect.

**Comment by rmehta** [comment]
@dottenbr What is the use case? Attachments to documents are already visible via the document view.

**Comment by gbayliss** [comment]
My company also has a requirement for this, with an added option for specifying the path per module.  This is so that we can integrate the attachments into our existing Windows shared folder infrastructure, although the same would apply to FOSS implementations with a very large number of files.

Files should be upload into a path which follows the form:

**<module.settings.path>/<doctype.name>/<document.name>/<attachment.filename>**

There are multiple reasons for  single biggest reason for this is to ensure that HR documentation cannot fall into the wrong hands (a requirement under Data Protection laws in the UK).

It would also be useful to revisit the random filename mechanism, which causes additional issues in this use case.  In order to maintain backward compatibility perhaps this could be enabled/disabled using an option in the settings.

**Comment by raveslave** [comment]
+1 
related request: https://github.com/frappe/erpnext/issues/8377

**Comment by stelden** [comment]
+1

Given that all attachments have a context, this is a must have feature. We can even have files and folders sync from a full-scale document manager and have them automatically picked as attachments by the relevant doctype. The possibilities are many.
